tests/integration/test.retry.js

```diff
@@ -0,0 +1,153 @@
+'use strict';
+
+var adapters = [
+  ['local', 'http'],
+  ['http', 'http'],
+  ['http', 'local'],
+  ['local', 'local']
+];
+
+if ('saucelabs' in testUtils.params()) {
+  adapters = [['local', 'http'], ['http', 'local']];
+}
+
+adapters.forEach(function (adapters) {
+  var suiteName = 'test.retry.js-' + adapters[0] + '-' + adapters[1];
+  describe(suiteName, function () {
+
+    var dbs = {};
+
+    beforeEach(function (done) {
+      dbs.name = testUtils.adapterUrl(adapters[0], 'testdb');
+      dbs.remote = testUtils.adapterUrl(adapters[1], 'test_repl_remote');
+      testUtils.cleanup([dbs.name, dbs.remote], done);
+    });
+
+    after(function (done) {
+      testUtils.cleanup([dbs.name, dbs.remote], done);
+    });
+    it('retry stuff', function (done) {
+
+      var remote = new PouchDB(dbs.remote);
+      var Promise = PouchDB.utils.Promise;
+      var allDocs = remote.allDocs;
+
+      // Reject attempting to write 'foo' 3 times, then let it succeed
+      var i = 0;
+      remote.allDocs = function (opts) {
+        if (opts.keys[0] === 'foo') {
+          if (++i !== 3) {
+            return Promise.reject(new Error('flunking you'));
+          }
+        }
+        return allDocs.apply(remote, arguments);
+      };
+
+      var db = new PouchDB(dbs.name);
+      var rep = db.replicate.from(remote, {
+        live: true,
+        retry: true
+      });
+
+      var paused = 0;
+      rep.on('paused', function (e) {
+        ++paused;
+        // The first paused event is the replication up to date
+        // and waiting on changes (no error)
+        if (paused === 1) {
+          (typeof e).should.equal('undefined');
+          return remote.put({}, 'foo').then(function () {
+            return remote.put({}, 'bar');
+          });
+        }
+        // Second paused event is due to failed writes, should
+        // have an error
+        if (paused === 2) {
+          e.should.exist();
+        }
+      });
+
+      var active = 0;
+      rep.on('active', function () {
+        ++active;
+      });
+
+      rep.on('complete', function () {
+        active.should.equal(4);
+        paused.should.be.at.least(3);
+        done();
+      });
+
+      rep.catch(done);
+
+      var numChanges = 0;
+      rep.on('change', function () {
+        if (++numChanges === 3) {
+          rep.cancel();
+        }
+      });
+
+      remote.put({}, 'hazaa');
+    });
+
+    it('source doesn\'t leak "destroyed" event', function (done) {
```

Comment:
this is the test that fails before the fix but succeeds after. In fact it just hangs rather than "failing," which is one of the annoying things I find about Mocha, but c'est la vie.